include/swift/SIL/SILSuccessor.h

```diff
@@ -87,33 +87,45 @@ class SILSuccessor {
   class pred_iterator {
     SILSuccessor *Cur;
 
+    // Cache the basic block to avoid repeated pointer chasing.
+    SILBasicBlock *BB;
+
+    void cacheBasicBlock();
+
   public:
     using difference_type = std::ptrdiff_t;
     using value_type = SILBasicBlock *;
     using pointer = SILBasicBlock **;
     using reference = SILBasicBlock *&;
     using iterator_category = std::forward_iterator_tag;
 
-    pred_iterator(SILSuccessor *Cur = 0) : Cur(Cur) {}
+    pred_iterator(SILSuccessor *Cur = nullptr) : Cur(Cur), BB(nullptr) {
+      cacheBasicBlock();
+    }
 
     bool operator==(pred_iterator I2) const { return Cur == I2.Cur; }
     bool operator!=(pred_iterator I2) const { return Cur != I2.Cur; }
 
     pred_iterator &operator++() {
-      assert(Cur && "Trying to advance past end");
+      // No need to assert(Cur). Dereferencing nullptr always crashes.
       Cur = Cur->Next;
+      cacheBasicBlock();
       return *this;
     }
 
-    pred_iterator operator++(int) {
-      pred_iterator copy = *this;
-      ++*this;
+    pred_iterator operator+(size_t distance) const {
```

Comment:
Do we really need to use a size_t here? I thought we were generally using int in such cases, no?

Reply:
Unless I missed something, the iterator is unidirectional, so an unsigned type would be appropriate here, right?

Reply:
I was just talking about the usage of size_t. We generally use unsigned for these cases.